File: uva_624.cpp
# include <iostream>
using namespace std;

int main(void){
    int n, t;
    int CD[30];
    int dp[1024][30];
    while(cin >> n >> t){
        for (int i=1; i<=t; ++i)
            cin >> CD[i];

        for (int i=0; i<1024; ++i)
            for(int j=0; j<30; ++j)
                dp[i][j]=0;
        for( int i = 1 ; i <= n ; i++ )
            for( int j = 1 ; j <= t ; j++ )
                dp[i][j] = (CD[j] <= i)?max(dp[i][j-1], dp[i-CD[j]][j-1]+CD[j]):dp[i][j-1];
        int _n=n, _t=t;
        while(_n && _t){
            if(dp[_n][_t] == dp[_n][_t-1])
                --_t;
            else{
                cout << CD[_t] << " ";
                _n-=CD[_t];
                --_t;
            }
        }
        cout << "sum:" << dp[n][t] << endl;
    }
    return 0;
}

File: uva_624.py
while True:
    try:
        input_s=input().split()
    except:
        break
    n, t=int(input_s[0]), int(input_s[1])
    CD=[0 for _ in input_s]
    for i in range(1, t+1):
        CD[i]=int(input_s[i+1])
    dp=[[0]*30 for _ in range(1024)]
    for i in range(1, n+1):
        for j in range(1, t+1):
            dp[i][j]=max(dp[i][j-1], dp[i-CD[j]][j-1]+CD[j]) if CD[j]<=i else dp[i][j-1]
    _n, _t=n, t
    while _n and _t:
        if dp[_n][_t]!=dp[_n][_t-1]:
            _n-=CD[_t]
            print("{} ".format(CD[_t]), end='')
        _t-=1
    print("sum:{}".format(dp[n][t]))